add initial design doc for Zed lake (#2500)
brimdata/zed: docs/lake/design.md

with the [Zed data model](../formats).

## Data Pools

A lake is comprised of _data pools_.  Each data pool is organized
according to its configured _pool key_.  Different data pools can have
different pool keys but all of the data in a pool must have the same
pool key.

The pool key is often a timestamp.  In this case, retention policies
and storage hierarchy decisions can be efficiently associated with
ranges of data over the pool key.

Data can be efficiently accessed via range scans comprised of a
range of values conforming to the pool key.

A lake has a configured sort order, either ascending or descending
and data is organized in the lake in accordance with this order.
Data scans may be either ascending or descending, and scans that
follow the configured order are generally more efficient than
scans that run in the opposing order.

Scans may also be range-limited but unordered.

If data loaded into a pool lacks the pool key, that data is still
imported but is not available to pool-key range scans.  Since it lacks
the pool key, such data is instead organized around its "." value.

> TBD: What is the interface for access non-keyed data?  Should this

> > show up in the zed language somehow?

Reviewer: ```suggestion
> show up in the Zed language somehow?
```

## Lake Semantics

The semantics of a Zed lake loosely follows the nomenclature and
design patterns of `git`.  In this approach,
* a _pool_ is like a `git` repository,
* a _load_ operation is like a `git add` followed by a `git commit`,
* and a lake _snapshot_ is like a `git checkout`.

A core theme of the Zed lake design is _ergonomics_.  Given the git metaphor,
our goal here is that the Zed lake tooling be as easy and familiar as git is
to a technical user.

While this design document is independent of any particular implementation,
we will illustrate examples of `zed lake` commands that are under development.

> Note that while this CLI-first approach provides an ergonomic way experiment with

Reviewer: ```suggestion
Note that while this CLI-first approach provides an ergonomic way to experiment with
```

> and learn the zed lake building blocks, all of this functionality is also

Reviewer: ```suggestion
and learn the Zed lake building blocks, all of this functionality is also
```

> exposed through an API to a cloud-base service.  Most interactions between

Reviewer: ```suggestion
exposed through an API to a cloud-based service.  Most interactions between
```

> a user and a Zed lake would be via an application like Brim or a

Reviewer: ```suggestion
a user and a Zed lake would be via an application like [Brim](https://github.com/brimdata/brim) or a
```

### Load

Data is then loaded into a lake with the `load` command, .e.g.,
```
zed lake load -p logs sample.ndjson
```
where `sample.ndjson` contains logs in NDJSON format.  Any supported format
(i.e., CSV, NDJSON, Parquet, ZNG, and ZST) as well multiple files can be used
here, e.g.,
```
zed lake load -p logs sample1.csv sample2.zng sample3.zng
```
Parquet and ZST formats are not auto-detected so you must currently specify
`-i` with these formats, e.g.,
```
zed lake load -p logs -i parquet sample4.parquet
zed lake load -p logs -i zst sample5.zst
```

Note that there is no need to define a schema or insert data into
a "table" as all Zed data is _self describing_ and can be queried in a
schema-agnostic fashion.  Data of any _shape_ can be stored in any pool
and arbitrary data _shapes_ can coexist side by side.

### Scan

Data is read from a pool with the `scan` command.  By default, `scan`
generates sorted output of all of the pool data in the configured key order.
The order can be overridden with `-order asc` or `-order desc`.

A range can be specified with `-from` or `-to`.

The default output format is ZNG though it can be overridden with the various
supported output formats.

> This example reads every record from the `logs` pool starting
> from time `2020-1-1T12:00`, in ascending order,
> send the  results as ZNG to stdout, then pipe the output to `zq` to count the record:

Reviewer: ```suggestion
This example reads every record from the `logs` pool starting
from time `2020-1-1T12:00`, in ascending order,
sends the  results as ZNG to stdout, then pipes the output to `zq` to count the records:
```

```

> zed lake scan -p logs -from 2020-1-1T12:00 -order asc | zq "count()" -

Reviewer: ```suggestion
zed lake scan -p logs -from 2020-1-1T12:00 -order asc | zq "count()"
```

Author: I'll add `-z` but don't understand dropping the stdin arg

Reviewer: Sorry, yeah, I hit the wrong one. I just added another review comment where I think the trailing `-` might have been a copy/paste mistake.

the lake.

Arbitrarily complex Zed queries can be executed over the lake in this fashion
and the planner can utilize cloud resources to parallelize and scale the
query over many parallel workers that simultaneously access the Zed lake data in
shared cloud storage (while also accessing locally cached copies of data).

### Add and Commit

Continuing the `git` metaphor, the Zed lake `load` operation is actually
decomposed into two steps: an `add` operation and a `commit` operation.
These steps can be explicitly run in stages, e.g.,
```
zed lake add -p logs sample.json
(<commit-tag> printed to stdout)
zed lake commit -p logs <commit-tag>
```
Likewise, you can stack multiple adds and commit them all at once, e.g., ,
```
zed lake add -p logs sample1.json
(<commit-tag-1> printed to stdout)
zed lake add -p logs sample2.parquet
(<commit-tag-2> printed to stdout)
zed lake add -p logs sample3.zng
(<commit-tag-3> printed to stdout)
zed lake commit -p logs <commit-tag-1>,<commit-tag-2>,<commit-tag-3>
```
The commit command also takes an optional title and message that is stored
in the commit journal for reference.  These messages are carried in

> a description record attached to every journel entry, which has a Zed

Reviewer: ```suggestion
a description record attached to every journal entry, which has a Zed
```

and non-overlapping.
Here, the segments from the given commit tag are read and compacted into
a new commit as an `add` operation.  Again, until the data is actually committed,
no readers will see any change.

Additionally, multiple commits can be merged all at once to sort all of the
segments of all of the commits that comprise the group, e.g.,
```
zed lake merge -p logs <commit-tag-1>,<commit-tag-2>,<commit-tag-3>
(<merged-commit-tag> printed to stdout)
```

### Squash

After the merge phase, we have a new commit that combines the old commits
across non-overlapping segments, but they are not yet committed.
To avoid consistency issues here, the old commits need
to be deleted while simultaneously adding the new commit.

This can be done with a form of the `commit` command
that specifies a list of
commits to drop in addition to the commit(s) to add, e.g.,
```
zed lake commit -p logs -drop <commit-tag-1>,<commit-tag-2>,<commit-tag-3> <new-commit-tag>
```

The `squash` command performs the merge and atomic commit as follows:
```
zed lake squash -p logs <commit-tag-1>,<commit-tag-2>,<commit-tag-3>
(<merged-commit-tag> printed to stdout)
```

> which is equivalent this sequence of commands:

Reviewer: ```suggestion
which is equivalent to this sequence of commands:
```

```
zed lake scan -p logs <commit-tag-1>,<commit-tag-2>,<commit-tag-3> | zed lake -p logs add -
(<new-commit-tag> printed to stdout)
zed lake commit -p logs -drop <commit-tag-1>,<commit-tag-2>,<commit-tag-3> <new-commit-tag>
```

> Also, note that Zed queries could be optionally applied to the `scan` operator
> providing a means to post-process data or delete certain fields based on

> > Zed filter syntax.  See below regarding "relationable tables".

Reviewer: ```suggestion
> Zed filter syntax.  See below regarding the [Relational Model](#relational-model).
```

policy over
how data is ingested and committed and rolled up.  Depending on the
use case and workflow, we envision that some amount of overlapping segments
would persist at small scale and always be "mixed in" with other overlapping
segments during any key-range scan.

> Note: since this style of data organization follows the LSM pattern,
> how data is rolled up (or not) can control the degree of LSM write
> amplification that occurs for a given workload.  There is an explicit
> tradeoff here between overhead of merging overlapping segments on read
> and LSM write amplification to organize the data to avoid such overlaps.

> Note: we are showing here manual, CLI-driven steps to accomplish these tasks
> but a live data pipeline would automate all of this with orchestration that
> performs these functions via a service API, i.e., the same service API
> used by the CLI operators.

### Log

Like `git log`, the command `zed lake log` prints the journal of commit
operations.

> TBD: define format of this output.  The info record context should be
> displayed along with the adds/drops of each commit tag.

## Cloud Object Naming

> A Zed lake semantics defined above are achieved by mapping the
> lake abstractions onto a key-value cloud obejct store.

Reviewer: ```suggestion
The Zed lake semantics defined above are achieved by mapping the
lake abstractions onto a key-value cloud object store.
```

Every data element in a Zed lake is one of three fundamental object types:
* a single-writer _immutable object_,
* a multi-writer _commit journal_, or
* a multi-writer _mutable object_.

### Immutable Objects

All data in a data pool is comprised of immutable objects, which are organized
into data _segments_.  Each segment is composed of one or more immutable objects
all of which share a common, globally unique identifier,
which refer to below as `<id>`.

These identifiers are [KSUIDs](https://github.com/segmentio/ksuid).
The KSUID allocation scheme
provides a decentralized solution for creating globally unique IDs.
KSUIDs have embedded timestamps so the creation time of
any object named in this way can be derived.  Also, a simple lexicographic
sort of the KSUIDs results in a creation-time ordering (though this ordering
is not relied on for causal relationships since clock skew can violate
such an assumption).

Segments are referred to by zero or more commits, where the commits
are maintained in a commit journal described below.

> > While a zed lake is defined in terms of a cloud object store, it may also

Reviewer: ```suggestion
> While a Zed lake is defined in terms of a cloud object store, it may also
```

> be realized on top of a file system, which provides a convenient means for
> local, small-scale deployments or test/debug workflows.  Thus, for simple use cases,
> the complexity of running an object-store service may be avoided.

#### Segments

An immutable object is created by a single writer using a globally unique name
with an embedded KSUID.  
New objects are written in their entirety.  No updates, appends, or modifications
may be made once an object exists.  Given these semantics, any such object may be
trivially cached as its name or content never changes.

> Since the writer the name is globally unique and the

Reviewer: I'm not sure how to best fix this, but "the writer the name" looks like mistaken wording.

|row seek index|`<id>-<segment>-seek.zng`|
|search index|`<id>-<segment>-index-<rule>.zng`|

`<id>` is the KSUID of the commit, whiel
the `<segment>` portion of the name is an integer representing the positional
index of the segment within that commit.  Segments are number from 0 to N-1,
where N segments define the commit.

The seek index maps pool key values to seek offset in the ZNG file thereby
allowing a scan to do a partial GET of the ZNG object when scanning only
a subset of data.

> Note the ZST format will have seekable checkpoints based on the sort key that
> are encoded into its metadata section so there is no need to have a separate
> seek index for the columnar object.

#### Search Indexes

To optimize pool scans, the lake design includes the well-known pruning
concept, where segments of data can be skipped when it can be determined
(either at "compile time" or "run time") that a segment of data is not
needed by a scan, e.g., because a filter predicate would otherwise filter
all of the data in that object.

In addition to standard techniques for pruning a cloud scan (e.g., summary stats
that can determine when a predicate would be false for every record, etc),

> search indexes can be built and attached to any data chunk using the

Reviewer: This is the first time the word "chunk" appears in the doc, and there's 5 instances of "chunk" alongside 50 instances of "segment". I _assume_ you were making a conscious attempt to transition away from the term "chunk" (which I support) but if I'm wrong about that and both concepts persist, it might help to explain the difference.

I'm not flagging every instance of "chunk" in these review comments. Just flagging this one to raise the topic.

Author: Yes, should be segment.  I'll do global replace.

is computed by applying the ADD/DROP directives in sequence from entry TAIL to
the journal entry in question, up to HEAD.  This gives the set of commit tags
that comprise a snapshot.  A snapshot may then be scanned by scanning,
in key order, the segments that comprise all of the commits while merging
records from overlapping segments.  The snapshot is sorted by its pool key
range, where key-range values are sorted by the beginning key as the primary key
and the ending key is the secondary key.

For efficiency, a journal entry's snapshot may be stored as "cached snapshot"
alongside the journal entry.  This way, the snapshot at HEAD may be
efficiently computed by locating the most recent cached snapshot and scanning
forward to HEAD.

#### Scaling the Journal

When the size of the snapshot file reaches a certain size (and thus becomes too large to
conveniently handle in memory), the journal is converted to an internal sub-pool
called the "journal pool".  The journal pool's
pool key is the "from" value (of its parent pool key) from each commit.
In this case, commits to the parent pool are made in the same fashion,

> but instead of snapshotting updates into a snapshot zng file,

Reviewer: ```suggestion
but instead of snapshotting updates into a snapshot ZNG file,
```

You've been pretty consistent in here about capitalizing ZNG when speaking about it in sentences (which I support), so, might as well keep it up. :smile:

Mutable objects can be modified by any writer and are stored in a directory
named after the configuration object.  Here, each entry represents the entire
mutable object as a ZNG file numbered from 0 upward, i.e., `0.zng`, `1.zng`,
and so forth.  The `HEAD` file points at the last and valid object in the sequence.
Once the HEAD has been atomically advanced, the previous mutable object can be deleted
(or the mutable objects can be preserved to keep a history of the configuration changes).

### Object Naming

```
lake-<lake-id>/
  config/
    HEAD
    0.zng
    1.zng
    2.zng
    ...
  index-rules/
    HEAD
    0.zng
    1.zng
    2.zng
    ...
  pool-<pool-id>/
    config/
      TAIL
      0.zng
      1.zng
      ...
    journal/
      HEAD
      TAIL
      0.zng
      1.zng
      2.zng
      ...
      20.zng
      20-snap.zng
      20-seek.zng
      21.zng
      ...
    data/
      <tag>-0-.{zng,zst}
      <tag>-0-seek.zng
      <tag>-0-index-0.zng
      <tag>-0-index-1.zng
      ...
      <tag>-1-.{zng,zst}
      <tag>-1-seek.zng
      <tag>-1-index-0.zng
      <tag>-1-index-1.zng
      ...
  pool-<pool-id>/
  ...
```

## Continuous Ingest

> While the description above is very batch oriented, the Zed like design is

Reviewer: ```suggestion
While the description above is very batch oriented, the Zed lake design is
```

intended to perform scalably for continuous streaming applications.  In this
approach, many small commits may be continuously executed as data arrives and
after each commit, the data is immediately readable.

> To handle this use case, the _journal_ of commits is design

Reviewer: ```suggestion
To handle this use case, the _journal_ of commits is designed
```

> TBD: define model for scanning metadata in this fashion.  It might be as
> easy as scanning virtual sub-pools that conform to the different types of
> metadata related to a pool, e.g., logs.$segments, logs.$commits, etc.

## Derived Analytics

To improve the performance of predictable workloads, many use cases of a
Zed lake pre-compute _derived analytics_ or a particular set of _partial
aggregations_.

For example, the Brim app displays a histogram of event counts grouped by
a category over time.  The partial aggregation for such a computation can be
configured to run automatically and store the result in a pool designed to
hold such results.  Then, when a scan is run, the Zed analytics engine
recognizes when the DAG of a query can be rewritten to assemble the
partial results instead of deriving the answers from scratch.

When and how such partial aggregations are performed is simply a matter of
writing Zed queries that take the raw data and produce the derived analytics
while conforming to a naming model that allows the Zed lake to recognize
the relationship between the raw data and the derived data.

> TBD: Work out these details which are reminiscent of the analytic cache

> > developed in our earlier prototype.  For example,

Reviewer: ```suggestion
> developed in our earlier prototype.
```

## Keyless Data

This is TBD.  Data without a key should be accepted some way or another.

> One approach could be to simply assign the "zero-value" as the pool key.
> Or a configured default value could be user.  This would make key-based

Reviewer: ```suggestion
One approach could be to simply assign the "zero-value" as the pool key,
or a configured default value could be used.  This would make key-based
```

retention policies more complicated.

Another approach would be to create a sub-pool on demand when the first
keyless data is encountered, e.g., `pool-name.$nokey` where the pool key
is configured to be ".".  This way, an app or user could query this pool
by this name to scan keyless data.

## Relational Model

Since a Zed lake can provide strong consistency, workflows that manipulate
data in a lake can utilize a model where updates are made to the data
in place.  Such updates involve creating new commits from the old data
where the new data is a modified form of the old data.  This provides
emulation of row-based updates and deletes.

> Is the pool-key is chosen to be "." for such a use case, then unique

Reviewer: ```suggestion
If the pool-key is chosen to be "." for such a use case, then unique
```

rows can be maintained by trivially detected duplicates (because any
duplicate row will be adjacent when sorted by ".").
so that dups are trivially detected.

Efficient upserts can be accomplished because each chunk is sorted by the
pool key.  Thus, an upsert can be sorted then merge-joined with each
overlapping segment.  Where segments produce changes and additions, they can
be forwarded to a streaming add operator and the list of modified chunks
accumulated.  At the end of the operation, then new commit(s) along with
the to-be-deleted segments can be added to the journal in a single atomic
operation.  A write conflict occurs if there are any another deletes to
the list of to-be-deleted segments.  When this occurs, the transaction can
simply be restarted.  To avoid inefficiency of many restarts, an upsert can
be partitioned into smaller chunks if the use case allows for it.

> TBD finish working through this use case, its requirements, and the
> mechanisms needed to implement it.  Write conflicts will need to be
> managed at a layer above the journal or the journal extended with the
> needed functionality.

## Thoughts on Version 2

> Ref counted chunks stored outside of pool.

Reviewer: ```suggestion
Ref counted chunks stored outside of a pool.
```

> Copy-on-write semantics so massive pool can be instantaneously copied then

Reviewer: ```suggestion
Copy-on-write semantics so massive pools can be instantaneously copied then
```

changed/updated under the new pool.

## Next steps...

To get going, we will simplify the above design then create issues
to incrementally add support for each needed area.  In this way, we can
leverage the current code base to get up and running quickly.

The initial prototype will be simplified as follows:
* no transactions, as we will presume a single writer

>     * config/meta files stored as single zng files

Reviewer: ```suggestion
    * config/meta files stored as single ZNG files
```

* journal stored as sequence of complete snapshots with HEAD pointing to
most recent entry (HEAD written after update)
* no recursive journal pool
* no columnar support
* no search indexes
* pool-key sorted scans only
* no keyless intake

As we work through the `zed lake` command set and API,

> we should strive to make the zed lake CLI-first commands one-to-one

Reviewer: ```suggestion
we should strive to make the Zed lake CLI-first commands one-to-one
```

```

Note that there is no need to define a schema or insert data into
a "table" as all Zed data is _self describing_ and can be queried in a
schema-agnostic fashion.  Data of any _shape_ can be stored in any pool
and arbitrary data _shapes_ can coexist side by side.

### Scan

Data is read from a pool with the `scan` command.  By default, `scan`
generates sorted output of all of the pool data in the configured key order.
The order can be overridden with `-order asc` or `-order desc`.

A range can be specified with `-from` or `-to`.

The default output format is ZNG though it can be overridden with the various
supported output formats.

This example reads every record from the `logs` pool starting
from time `2020-1-1T12:00`, in ascending order,
sends the  results as ZNG to stdout, then pipes the output to `zq` to count the records:
```
zed lake scan -p logs -from 2020-1-1T12:00 -order asc | zq -z "count()" -
```

### Query

Of course, the example is an inefficient way to count records.  Instead of
reading the records out of the lake and processing them, a better approach
is to push the query into the lake.

The `query` command lets you do this, e.g.,
```

> zed lake query -p logs -from 2020-1-1T12:00 -order asc "count()" -

Reviewer: ```suggestion
zed lake query -p logs -from 2020-1-1T12:00 -order asc "count()"
```

Oops. I screwed up with my other review comment. This was the place where I intended to flag what I thought was a misplaced trailing `-`.
